src/System.IO.FileSystem/src/System/IO/Enumeration/EnumerationOptions.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.IO.Enumeration
+{
+    public class EnumerationOptions
+    {
+        /// <summary>
+        /// For internal use. These are the options we want to use if calling the existing Directory/File APIs where you don't
+        /// explicitly specify EnumerationOptions.
+        /// </summary>
+        internal static EnumerationOptions Compatible => new EnumerationOptions { MatchType = MatchType.Dos };
+
+        /// <summary>
+        /// Internal singleton for default options.
+        /// </summary>
+        internal static EnumerationOptions Default => new EnumerationOptions();
+
+        /// <summary>
+        /// Default constructor. Constructs the options class with recommended default options.
+        /// </summary>
+        public EnumerationOptions()
+        {
+        }
+
+        /// <summary>
+        /// Converts SearchOptions to FindOptions. Throws if undefined SearchOption.
+        /// </summary>
+        internal static EnumerationOptions FromSearchOption(SearchOption searchOption)
+        {
+            if ((searchOption != SearchOption.TopDirectoryOnly) && (searchOption != SearchOption.AllDirectories))
+                throw new ArgumentOutOfRangeException(nameof(searchOption), SR.ArgumentOutOfRange_Enum);
+
+            return searchOption == SearchOption.AllDirectories ? new EnumerationOptions { RecurseSubdirectories = true, MatchType = MatchType.Dos } : Compatible;
+        }
+
+        /// <summary>
+        /// Should we recurse into subdirectories while enumerating?
+        /// </summary>
+        public bool RecurseSubdirectories { get; set; }
+
+        /// <summary>
+        /// Skip files/directories when access is denied (e.g. AccessDeniedException/SecurityException)
+        /// </summary>
+        public bool IgnoreInaccessible { get; set; }
+
+        /// <summary>
+        /// Suggested buffer size.
+        /// </summary>
+        /// <remarks>
+        /// Not all platforms use user allocated buffers, and some require either fixed buffers or a
+        /// buffer that has enough space to return a full result. One scenario where this option is
+        /// useful is with remote share enumeration on Windows. Having a large buffer may result in
+        /// better performance as more results can be batched over the wire.
+        /// </remarks>
+        public int BufferSize { get; set; }
```

Comment:
Rather than taking a buffer size, should we instead have this be a `Memory<byte>` that would allow the caller to provide their own scratch buffer, such that it could be reused from iteration to iteration (and if we ever need to pin it, they could prepin it)?

Reply:
We do have to pin. I thought about how to approach doing this, but tabled it for future as it is a little strange and would require some discussion. We don't always use a buffer, and when we do they have to be a very specific minimum size. The platform-to-platform variance is pretty big, and could conceivably (even likely) change in the future too.

Reply:
How does a developer know what to set this to?

Reply:
It is advisory and requires pretty intimate knowledge of how things work. It is essentially the equivalent of FIND_FIRST_EX_LARGE_FETCH in FindFirstFileEx which people have been asking for. I started with `UseLargeBuffer` but feedback was given in API review to change it to a specific number.

I'll add some more comments to try and help clarify. Note that this option is being specifically requested by customers (not in this shape exactly- they want the FIND_FIRST_EX_LARGE_FETCH option).